File: 000001-000200/000049.字母异位词分组/49.js
/**
 * @param {string[]} strs
 * @return {string[][]}
 */
// 方法一：用时 148 ms
// 执行用时超过了 27 %
// 消耗内存超过了 81 %
// 难度: 中等
var groupAnagrams = function(strs) {
    if (strs.length == 0) {
        return strs
    }
    if (strs.length == 1) {
        return [strs]
    }
    let ret = new Map()
    for (const item of strs) {
        let key = Array.from(item).sort().toString()
        let list = ret.get(key) ? ret.get(key) : new Array()
        list.push(item)
        ret.set(key, list)
    }
    return Array.from(ret.values())
};

// 方法二：用时 112 ms
// 执行用时超过了 88 %
// 消耗内存超过了 77 %
// 难度: 简单
var groupAnagrams = function(strs) {
    if (strs.length == 0) {
        return strs
    }
    if (strs.length == 1) {
        return [strs]
    }
    let ret = new Map()
    for (const item of strs) {
        let key = item.split("").sort().join("")
        ret.get(key) ? ret.get(key).push(item) : ret.set(key, [item])
    }
    return Array.from(ret.values())
};

console.log(groupAnagrams(["eat", "tea", "tan", "ate", "nat", "bat"]))
console.log(groupAnagrams([""]))
console.log(groupAnagrams(["a"]))

File: 000001-000200/000049.字母异位词分组/49.go
package main

import (
	"fmt"
	"sort"
)

// 方法一：用时 20 ms
// 执行用时超过了 89 %
// 消耗内存超过了 69 %
// 难度: 中等
func groupAnagrams(strs []string) [][]string {
	smap := map[string][]string{}
	for _, str := range strs {
		s := []byte(str)
		sort.Slice(s, func(i, j int) bool { return s[i] < s[j] })
		key := string(s)
		smap[key] = append(smap[key], str)
	}
	ret := make([][]string, 0, len(smap))
	for _, v := range smap {
		ret = append(ret, v)
	}
	return ret
}

func main ()  {
	fmt.Println(groupAnagrams([]string{"eat", "tea", "tan", "ate", "nat", "bat"}))
	fmt.Println(groupAnagrams([]string{""}))
	fmt.Println(groupAnagrams([]string{"a"}))
}